domain/access/modelmigration/import.go

```diff
@@ -0,0 +1,109 @@
+type Coordinator interface {
+	// Add adds the given operation to the migration.
+	Add(modelmigration.Operation)
+}
+
+// RegisterImport registers the import operations with the given coordinator.
+func RegisterImport(coordinator Coordinator, logger logger.Logger) {
+	coordinator.Add(&importOperation{
+		logger: logger,
+	})
+}
+
+// ImportService provides a subset of the access domain
+// service methods needed for model permissions import.
+type ImportService interface {
+	// CreatePermission gives the user access per the provided spec.
+	// If the user provided does not exist or is marked removed,
+	// accesserrors.PermissionNotFound is returned.
+	// If the user provided exists but is marked disabled,
+	// accesserrors.UserAuthenticationDisabled is returned.
+	// If a permission for the user and target key already exists,
+	// accesserrors.PermissionAlreadyExists is returned.
+	CreatePermission(ctx context.Context, spec corepermission.UserAccessSpec) (corepermission.UserAccess, error)
+	// SetLastModelLogin will set the last login time for the user to the given
+	// value. The following error types are possible from this function: -
+	// accesserrors.UserNameNotValid: When the username supplied is not valid. -
+	// accesserrors.UserNotFound: When the user cannot be found. -
+	// modelerrors.NotFound: If no model by the given modelUUID exists.
+	SetLastModelLogin(ctx context.Context, name user.Name, modelUUID coremodel.UUID, time time.Time) error
+}
+
+type importOperation struct {
+	modelmigration.BaseOperation
+
+	logger  logger.Logger
+	service ImportService
+}
+
+// Name returns the name of this operation.
+func (i *importOperation) Name() string {
+	return "import model user permissions"
+}
+
+// Setup implements Operation.
+func (i *importOperation) Setup(scope modelmigration.Scope) error {
+	i.service = service.NewService(
+		state.NewState(scope.ControllerDB(), i.logger))
+	return nil
+}
+
+// Execute the import on the model user permissions contained in the model.
+func (i *importOperation) Execute(ctx context.Context, model description.Model) error {
+	modelUUID := model.Tag().Id()
+	for _, u := range model.Users() {
+		name := user.NameFromTag(u.Name())
+		access := corepermission.Access(u.Access())
+		if err := access.Validate(); err != nil {
+			return errors.Annotatef(err, "importing access for user %q", name)
+		}
+		_, err := i.service.CreatePermission(ctx, corepermission.UserAccessSpec{
+			AccessSpec: corepermission.AccessSpec{
+				Target: corepermission.ID{
+					ObjectType: corepermission.Model,
+					Key:        modelUUID,
+				},
+				Access: access,
+			},
+			User: name,
+		})
+		if err != nil && !errors.Is(err, accesserrors.PermissionAlreadyExists) {
+			// If the permission already exists then it must be the model owner
+			// who is granted admin access when the model is created.
+			return errors.Annotatef(err, "creating permission for user %q", name)
```

Comment:
**issue(probable):** This particular error shouldn't cause the whole import to fails, because we are talking about an error of an imported user which has already permission on it. In a loop seems overkill to stop the loop for this kind of error isn't it ?

Reply:
I would rather not ignore the error unconditionally, as it is we are ignoring it only if it is `PermissionAlreadyExists`, but it could be something like "cannot connect to database" in which case we should be reporting it back here.